### PR TITLE
[General] Configuration guidelines

### PR DESCRIPTION
**Pull request by bterlson**
For discussion during today's meeting, and based on discussion from last time and further side conversations in the meantime.

### CODE REVIEW

File: docs/design/Configuration.mdk [hunk 1]
```diff
@@ -1,9 +1,232 @@
...
+~ Must
+allow clients to opt out of all global configuration settings at once
+~
+
+~ Must
+allow all global configuration settings to be overridden by client-provided options.
+~
+
+~ MustNot
+change behavior based on configuration changes that occur after the client is constructed. The only exceptions to this are the logging and HTTP Agent singletons.
```
Comment (adrianhall) [inline docs/design/Configuration.mdk]: I'd question the validity of allowing exceptions generally (maybe in an idiomatic language-specific version?)

**Comment by johanste** [comment]
Should we include [.netrc](https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html) in the list of configuration options?

**Comment by bterlson** [comment]
Everything is updated based on our previous discussions. Please review and give any further feedback you might have.

Here's a couple items we deferred and so are not mentioned in these guidelines:

* Requirements around Trusted Certificate Store (Node doesn't use it, but does support NODE_EXTRA_CA_CERTS environment variable).
* AZURE_LOG_LEVEL and its appropriateness as a general guideline.

File: docs/design/Configuration.mdk [hunk 2]
```diff
@@ -1,9 +1,130 @@
...
+~
+
+~ Must
+support CIDR notation for NO_PROXY
+~
+
+#### Client-library Specific Environment Variables
+
+~ May
+use client-libary-specific environment variables for portal-configured settings which are provided as parameters to your SDK. This generally includes credentials and connection details. For example, Service Bus could support the following environment variables:
```
Comment (JeffreyRichter) [inline docs/design/Configuration.mdk]: SDK --> client library

File: docs/design/Configuration.mdk [hunk 3]
```diff
@@ -1,9 +1,130 @@
...
+~
+
+~ Must
+get approval for every new environment variable. 
+~
+
+~ Must
+use this syntax for environment variables specific to a particular Azure service:
+
+`AZURE_<ServiceName>_<ConfigurationKey>`
```
Comment (JeffreyRichter) [inline docs/design/Configuration.mdk]: Somewhere you should say that all uppercase letters are required.

File: docs/design/Configuration.mdk [hunk 4]
```diff
@@ -1,9 +1,130 @@
...
+
+### Global Configuration
+
+Global configuration refers to configuration settings that are applied to all applicable client constructors in some manner. 
+
+~ Must
+support global configuration of shared pipeline policies including:
+
+* Logging: Log level, swapping out logger implementation
+* HTTP: Proxy settings, max retries, timeout, swapping out Agent implementation
```
Comment (JeffreyRichter) [inline docs/design/Configuration.mdk]: I thought we didn't want retries and I'm not sure what "timeout" means?
If it's about retries, then there can be multiple timeouts: per-try timeout, per-operation timeout, max retry delay timeout, timeout between tries, etc.

File: docs/design/Configuration.mdk [hunk 5]
```diff
@@ -1,9 +1,130 @@
...
+~ Must
+use relevant global configuration settings either by default or when explicitly requested to by the user e.g. by passing in a configuration object to a client constructor.
+~
+
+~ Must
+allow different clients of the same type to use different configurations.
+~
+
+~ Must
+allow consumers of service clients to opt out of all global configuration settings at once, both via global "off switch" and a per-client opt-out.
```
Comment (JeffreyRichter) [inline docs/design/Configuration.mdk]: I think you mean programmatically? 
I can't opt-out via another configuration key or setting, right?

File: docs/design/Configuration.mdk [hunk 6]
```diff
@@ -1,9 +1,130 @@
...
+~
+
+~ Must
+allow all global configuration settings to be overridden by client-provided options. The names of these options should align with any user-facing global configuration keys.
```
Comment (JeffreyRichter) [inline docs/design/Configuration.mdk]: Do you mean programmatically in code or do you mean by passing config values to a config system in code in Main (or language equivalent)?

File: docs/design/EnvironmentVariables.mdk [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+| AZURE_CLIENT_ID            | AAD                                    |
+| AZURE_CLIENT_SECRET        | AAD                                    |
+| AZURE_TENANT_ID            | AAD                                    |
+| AZURE_RESOURCE_GROUP       | Azure RG                               |
+| AZURE_CLOUD                | mooncake, govcloud, etc.               |
+| **Pipeline Configuration**                                         ||
+| AZURE_TELEMETRY_DISABLED   | Disables telemetry                     |
+| AZURE_LOG_LEVEL            | Enable console logging by setting a log level. |
+| AZURE_TRACING_DISABLED     | Disables tracing                       |
+~
```
Comment (JeffreyRichter) [inline docs/design/EnvironmentVariables.mdk]: These look good to me. I don't see retry (which I like) so it should be removed from up above.

File: docs/design/Configuration.mdk [hunk 8]
```diff
@@ -1,9 +1,130 @@
...
+2. Environment variables
+3. Global configuration store (code)
+4. Runtime parameters
+
+~ Must
+apply configuration in the order above by default, such that subsequent items in the list override settings from previous items in the list.
+~
+
+~ May
+support configuration systems that users opt in to that do not follow the above ordering.
```
Comment (JonathanGiles) [inline docs/design/Configuration.mdk]: I can't parse this sentence, particularly the "that do not follow the above ordering" part.

Comment (JonathanGiles) [reply docs/design/Configuration.mdk]: Do you just mean that "you can support other configuration systems"?

Comment (bterlson) [reply docs/design/Configuration.mdk]: Yeah basically. The wording is confusing, I will clean up after this PR.

**Pull request merged**